### PR TITLE
[Statistics][Instrument] Added permission check for viewing per instrument info.

### PR DESCRIPTION
**Pull request by AnyhowStep**
In reference to, #2772 

I just added a permission check. I'm not totally sure if these are the correct permissions. Can someone take a look?

### CODE REVIEW

**Comment by AnyhowStep** [comment]
After double checking, the problem isn't with the statistics module. It's a problem with the instrument thingy.

**Comment by AnyhowStep** [comment]
Well, minor problem with the statistics module. It should really have "data entry" permission checks.

**Comment by AnyhowStep** [comment]
In reference to Redmine ticket, 12324

This pull request also fixes a security bug where a user can get write-access to an instrument they're not supposed to.

**Comment by AnyhowStep** [comment]
This is not being rebased to 17.0 because the fixes that must be applied to both versions is different; because of multi-site.

**Comment by AnyhowStep** [comment]
@kongtiaowang I'm getting cryptic PHP Unit errors I'm not skilled enough to identify the problem behind. Does PHP Unit have problems with writing to PHP superglobals like `$_REQUEST`?

**Comment by AnyhowStep** [comment]
It finally passed Travis. The problem here is that LorisForm tries to use `$_POST` even though the `REQUEST_METHOD` is a `GET`. Also, some security exploits regarding instruments are getting patched.

File: php/libraries/User.class.inc [hunk 1]
```diff
@@ -411,5 +411,37 @@ class User extends UserPermissions
...
+    /**
+     * Determines if the user has a permission
+     * for a center
+     *
+     * @param string $code      The permission code
+     * @param int    $center_id The center id
+     *
+     * @return bool
+     */
+    public function hasCenterPermission($code, $center_id)
```
Comment (AnyhowStep) [inline php/libraries/User.class.inc]: Added so we don't check center-specific permissions incorrectly.

File: php/libraries/User.class.inc [hunk 2]
```diff
@@ -411,5 +411,37 @@ class User extends UserPermissions
...
     }
+
+    /**
+     * Determines if the user has a center
+     *
+     * @param int $center_id The center id
+     *
+     * @return bool
+     */
+    public function hasCenter($center_id)
```
Comment (AnyhowStep) [inline php/libraries/User.class.inc]: Added to reduce boilerplate, which I've noticed littered around the code base.

File: modules/statistics/php/NDB_Menu_statistics_site.class.inc [hunk 3]
```diff
@@ -23,7 +23,16 @@
  */
 class NDB_Menu_Statistics_Site extends NDB_Menu
 {
-
+    /**
+     * Checking user's permission
+     *
+     * @return bool
+     */
+    function _hasAccess()
```
Comment (AnyhowStep) [inline modules/statistics/php/NDB_Menu_statistics_site.class.inc]: Added to be consistent with `NDB_Form_statistics.class.inc`

**Comment by AnyhowStep** [comment]
`NDB_Caller` was calling `$instrument->save()` **BEFORE** checking for write-access. Lol.

File: php/libraries/User.class.inc (hunk 1 above)
Comment (AnyhowStep) [inline php/libraries/User.class.inc]: An example of checking incorrectly: https://github.com/aces/Loris/blob/17.1-dev/modules/timepoint_list/php/TimePoint_List_ControlPanel.class.inc#L65

Comment (ridz1208) [inline php/libraries/User.class.inc]: @AnyhowStep there are no center-specific permissions yet

File: php/libraries/NDB_Caller.class.inc [hunk 4]
```diff
@@ -443,32 +443,37 @@ class NDB_Caller
...
+        $session_id = $instrument->getSessionID();
+        $timepoint  = TimePoint::singleton($session_id);
+        if (is_null($timepoint)) {
+            throw new Exception("Time point is null for {$session_id}");
+        }
+        //This preserves backwards compatability for any
+        //code that will be called that assumes
+        //$_REQUEST["sessionID"] will be populated
+        $_REQUEST["sessionID"] = $session_id;
+        if (strtoupper($_SERVER["REQUEST_METHOD"]) === "GET") {
```
Comment (ridz1208) [inline php/libraries/NDB_Caller.class.inc]: @AnyhowStep why was this added ? where it it used ?

**Comment by ridz1208** [comment]
This might affect a few different places in the code, I added @driusan to the reviewers to evaluate the risk

File: php/libraries/User.class.inc (hunk 1 above)
Comment (AnyhowStep) [inline php/libraries/User.class.inc]: The point is to have code start assuming there is, right now. Our current ad-hoc logic, littered all over the code base, does this, anyway. But they're checking it very differently.

Comment (AnyhowStep) [inline php/libraries/User.class.inc]: This method is so all those different places start checking for it in the same way and, when we finally implement our magical center-specific permissions, we can modify the implementation of this method without changing all the other spots.

**Comment by driusan** [comment]
Please split this up into multiple pull requests that each only do one thing.

The statistics permission check looks good, the userpermissions functionality looks fine too (though I agree with @ridz1208 that it should be used), but the changes to NDB_Caller look highly risky and has changes unrelated to the pull request description which says it's "just adding a permission check" that would need to be evaluated independently.

**Comment by AnyhowStep** [comment]
The permission check was added to `NDB_Caller` for instruments. Hence, the title. All those changes to `NDB_Caller` were exactly for just permission checks.

**Comment by AnyhowStep** [comment]
It's not like I modified `NDB_Caller` and added a crazy new functionality that isn't related to just permission checks. I said I added permission checks and that was the only thing I did.

And the change to `NDB_Caller` is using the change to `User`. It *is* being used.
https://github.com/aces/Loris/pull/2774/files#diff-889d2c1aba1c7ce86bcab066ab79d21bR499

And statistics has that check because this pull request is part of redmine 12324, where people can access instruments from statistics.

**Comment by driusan** [comment]
Please do not reopen pull requests that were closed without incorporating the changes asked for.

**Comment by AnyhowStep** [comment]
@johnsaigle I'm leaving soon-ish. About a month. But it would be nice to see this security hole fixed. I'm pinging you as the best guy to tackle this =)

**Pull request closed without merging**